src/lib.rs

```diff
@@ -4,4 +4,568 @@
+use std::fs::File;
+use std::io::Read;
+use std::net::IpAddr;
+use std::net::Ipv4Addr;
+use std::str::FromStr;
+use std::sync::mpsc::Sender;
+use std::thread;
+use std::vec::Vec;
+
+use state_machine_future::{transition, RentToOwn, StateMachineFuture};
+use tokio::prelude::*;
+use trust_dns::client::{Client, SyncClient};
+use trust_dns::op::ResponseCode;
+use trust_dns::rr::{DNSClass, Name, RData, RecordType};
+use trust_dns::udp::UdpClientConnection;
+
+use cueball::backend;
+
+// TODO:
+// - checking based on A record TTLs
+// - retries
+// - tests
+// - shutdown
+// - last error
+// - comments
+// - unused (but used) import errors
+
+pub trait Resolver: Send + 'static {
+    fn start(&mut self, s: Sender<BackendMsg>);
+    fn stop(&mut self);
+    fn get_last_error(&self) -> Option<String>;
+}
+
+#[derive(Debug)]
+pub struct BackendAddedMsg {
+    pub key: backend::BackendKey,
+    pub backend: backend::Backend,
+}
+
+#[derive(Debug)]
+pub struct BackendRemovedMsg(pub backend::BackendKey);
+
+#[derive(Debug)]
+pub enum BackendMsg {
+    AddedMsg(BackendAddedMsg),
+    RemovedMsg(BackendRemovedMsg),
+}
+
+pub enum BackendAction {
+    BackendAdded,
+    BackendRemoved,
+}
+
+#[derive(Debug)]
+pub struct CueballResolver {
+    resolvers: Vec<String>,
+    domain: String,
+    service: String,
+    default_port: u16,
+    log: Logger,
+}
+
+impl CueballResolver {
+    pub fn new(
+        resolvers: &Vec<String>,
+        domain: String,
+        service: String,
+        port: u16,
+        log: Logger,
+    ) -> Self {
+        CueballResolver {
+            resolvers: resolvers.to_vec(),
+            domain,
+            service,
+            default_port: port,
+            log: log.clone(),
+        }
+    }
+}
+
+impl Resolver for CueballResolver {
+    fn start(&mut self, s: Sender<BackendMsg>) {
+        let srv_retry_params = RetryParams {
+            max: 0,
+            count: 0,
+            timeout: 1000,
+            min_delay: 0,
+            delay: 1000,
+            max_delay: 10000,
+        };
+
+        let srv_rec = SrvRec {
+            name: String::new(),
+            port: 0,
+            addresses_v4: Vec::new(),
+            expiry_v4: None,
+        };
+
+        let resolver = ResolverFSM::start(ResolverContext {
+            resolvers: self.resolvers.clone(),
+            backends: HashMap::new(),
+            srv: srv_rec,
+            srvs: Vec::new(),
+            srv_rem: Vec::new(),
+            service: self.service.clone(),
+            domain: self.domain.clone(),
+            last_srv_ttl: None,
+            next_service: None,
+            srv_retry_params: srv_retry_params,
+            dns_client: None,
+            defport: Some(80),
+            pool_tx: s,
+            log: self.log.clone(),
+        });
+
+        let _fsm = thread::spawn(move || {
+            let _result = resolver.wait();
+        });
+    }
+
+    fn stop(&mut self) {
+        std::unimplemented!()
```

Comment:
I'm curious what this implementation will look like. Will you be able to initiate an FSM transition to the Stop state from this thread?

Reply:
That was the idea, though I am fuzzy on the implementation right now. Were you looking for something different?

Reply:
No, I don't have enough familiarity with the details yet to have an opinion on how it should be done. It was really just curiosity that made me comment after a first look at the changes.